syne_tune/optimizer/schedulers/median_stopping_rule.py

```diff
@@ -0,0 +1,98 @@
+"""
+Example showing how to implement a new Scheduler.
+"""
+import logging
+from collections import defaultdict
+from typing import Optional, Dict, List
+
+import numpy as np
+
+from syne_tune.backend.trial_status import Trial
+from syne_tune.optimizer.scheduler import TrialScheduler, SchedulerDecision, TrialSuggestion
+
+
+class MedianStoppingRule(TrialScheduler):
+    def __init__(
+            self,
+            scheduler: TrialScheduler,
+            resource_attr: str,
+            running_average: bool = True,
+            metric: Optional[str] = None,
+            grace_time: Optional[int] = 1,
+            grace_population: int = 5,
+            rank_cutoff: float = 0.5,
+    ):
+        """
+        Applies median stopping rule in top of an existing scheduler.
+        * If result at time-step ranks less than the cutoff of other results observed at this time-step, the trial is
+        interrupted and otherwise, the wrapped scheduler is called to make the stopping decision.
+        * Suggest decisions are left to the wrapped scheduler.
+        * The mode of the wrapped scheduler is used.
+        Reference: Google Vizier: A Service for Black-Box Optimization. Golovin et al. 2017.
+        :param scheduler: scheduler to be called for trial suggestion or when median-stopping-rule decision is to
+        continue.
+        :param resource_attr: key in the reported dictionary that accounts for the resource (e.g. epoch or
+        wall-clocktime).
+        :param running_average: if True, then uses the running average of observation instead of raw observations.
+        :param metric: metric to be considered.
+        :param grace_time: median stopping rule is only applied for results whose `time_attr` exceeds this amount.
+        :param grace_population: median stopping rule when at least `grace_population` have been observed at a resource
+        level.
+        :param rank_cutoff: results whose quantiles are bellow this level are discarded (discard by default trials
+         whose results are bellow the median).
+        """
+        super(MedianStoppingRule, self).__init__(config_space=scheduler.config_space)
+        self.metric = scheduler.metric if metric is None else metric
+        self.sorted_results = defaultdict(list)
+        self.scheduler = scheduler
+        self.resource_attr = resource_attr
+        self.rank_cutoff = rank_cutoff
+        self.grace_time = grace_time
+        self.min_samples_required = grace_population
+        self.running_average = running_average
+        if running_average:
```

Comment:
Wouldn't it make sense to use an exponential moving average instead?

Reply:
It could but it would be a step from the paper and also introduce another parameter.